File: apps/android/app/src/main/java/app/bananatape/mobile/editor/ProviderPipelineState.kt
package app.bananatape.mobile.editor

import app.bananatape.mobile.adapters.AdapterError
import app.bananatape.mobile.adapters.NetworkReachability
import java.nio.file.Path

enum class MockProviderScenario {
    SUCCESS,
    SLOW_SUCCESS,
    PROVIDER_ERROR,
}

enum class AppLifecyclePhase {
    FOREGROUND,
    BACKGROUND,
}

data class ProviderReferenceImage(
    val id: String,
    val assetPath: String,
)

data class ProviderRequest(
    val id: String,
    val prompt: String,
    val provider: EditorProvider,
    val mode: EditorMode,
    val outputSize: OutputSize,
    val parentImageId: String?,
    val parentHistoryId: String?,
    val annotations: CanvasAnnotations,
    val references: List<ProviderReferenceImage>,
    val inputImagePath: Path? = null,
    val maskImagePath: Path? = null,
)

data class ProviderImageResult(
    val requestId: String,
    val imageUrl: String,
    val assetId: String,
    val assetPath: String,
    val size: EditorSize,
    val createdAt: String,
    val timestamp: Double,
)

sealed interface MockProviderResult {
    data class Success(val value: ProviderImageResult) : MockProviderResult
    data class Failure(val message: String) : MockProviderResult
}

data class MockImageProvider(
    val scenario: MockProviderScenario = MockProviderScenario.SUCCESS,
) {
    fun generate(request: ProviderRequest): MockProviderResult = resolve(request)
    fun edit(request: ProviderRequest): MockProviderResult = resolve(request)

    private fun resolve(request: ProviderRequest): MockProviderResult = when (scenario) {
        MockProviderScenario.SUCCESS,
        MockProviderScenario.SLOW_SUCCESS,
        -> {
            val suffix = stableSuffix(request)
            MockProviderResult.Success(
                ProviderImageResult(
                    requestId = request.id,
                    imageUrl = "mock://images/$suffix.png",
                    assetId = "asset-$suffix",
                    assetPath = "assets/$suffix.png",
                    size = EditorSize(width = 1024.0, height = 1024.0),
                    createdAt = "1970-01-01T00:00:00.000Z",
                    timestamp = if (request.mode == EditorMode.GENERATE) 1.0 else 2.0,
                ),
            )
        }
        MockProviderScenario.PROVIDER_ERROR -> MockProviderResult.Failure(ErrorMessage)
    }

    private fun stableSuffix(request: ProviderRequest): String {
        val normalizedPrompt = request.prompt
            .lowercase()
            .split(Regex("[^a-z0-9]+"))
            .filter { it.isNotBlank() }
            .take(6)
            .joinToString("-")
            .ifBlank { "annotation" }
        val mode = if (request.mode == EditorMode.GENERATE) "generate" else "edit"
        return "mock-$mode-$normalizedPrompt"
    }

    companion object {
        const val ErrorMessage = "Mock provider failed. Try again."
    }
}

data class ProviderPipelineState(
    val images: List<CanvasImage> = emptyList(),
    val history: List<HistoryEntry> = emptyList(),
    val focusedImageId: String? = null,
    val activeRequestId: String? = null,
    val activeParentHistoryId: String? = null,
    val activePreviousFocusedImageId: String? = null,
    val userErrorMessage: String? = null,
    val lifecyclePhase: AppLifecyclePhase = AppLifecyclePhase.FOREGROUND,
) {
    val pendingImages: List<CanvasImage> = images.filter { it.status == ImageGenerationStatus.PENDING }
    val readyImages: List<CanvasImage> = images.filter { it.status == ImageGenerationStatus.READY }
    val focusedImage: CanvasImage? = images.firstOrNull { it.id == focusedImageId }
    val focusedAnnotations: CanvasAnnotations = focusedImage?.annotations ?: CanvasAnnotations.Empty
    val focusedHistoryEntry: HistoryEntry? = focusedImage?.let { image -> history.firstOrNull { it.id == image.id || it.assetId == image.assetId } }
    val historyBrowserState: NativeHistoryBrowserState = NativeHistoryBrowserState(entries = history, selectedEntryId = focusedHistoryEntry?.id)
    val focusedLineage: FocusedImageLineage = focusedImageLineage(images, focusedImageId)

    fun startingGenerate(
        prompt: String,
        requestId: String,
        network: NetworkReachability,
        provider: EditorProvider = EditorProvider.MOCK,
    ): ProviderPipelineState {
        if (network == NetworkReachability.OFFLINE) return withError(AdapterError.Offline.userMessage)
        val placeholder = placeholderImage(
            id = pendingImageId(requestId),
            requestId = requestId,
            prompt = prompt,
            provider = provider,
            mode = EditorMode.GENERATE,
            parentId = null,
            generationIndex = images.size,
            generationBatchId = requestId,
            batchIndex = 0,
        )
        return copy(
            images = images + placeholder,
            focusedImageId = this.focusedImageId,
            activeRequestId = requestId,
            activeParentHistoryId = null,
            activePreviousFocusedImageId = focusedReadyImage()?.id,
            userErrorMessage = null,
        )
    }

    fun startingEdit(
        prompt: String,
        annotations: CanvasAnnotations,
        requestId: String,
        network: NetworkReachability,
        provider: EditorProvider = EditorProvider.MOCK,
    ): ProviderPipelineState {
        if (network == NetworkReachability.OFFLINE) return withError(AdapterError.Offline.userMessage)
        val parentImage = focusedReadyImage() ?: return withError("Select an image before editing.")
        val parentHistory = history.lastOrNull { it.id == parentImage.id || it.assetId == parentImage.assetId } ?: return withError("Select an image before editing.")
        val placeholder = placeholderImage(
            id = pendingImageId(requestId),
            requestId = requestId,
            prompt = prompt,
            provider = provider,
            mode = EditorMode.EDIT,
            parentId = parentImage.id,
            generationIndex = images.size,
            annotations = annotations,
            generationBatchId = requestId,
            batchIndex = 0,
        )
        return copy(
            images = images + placeholder,
            focusedImageId = this.focusedImageId,
            activeRequestId = requestId,
            activeParentHistoryId = parentHistory.id,
            activePreviousFocusedImageId = parentImage.id,
            userErrorMessage = null,
        )
    }

    fun applying(result: ProviderImageResult): ProviderPipelineState {
        if (activeRequestId != result.requestId) return this
        val pending = images.firstOrNull { it.id == pendingImageId(result.requestId) && it.status == ImageGenerationStatus.PENDING } ?: return this
        val ready = pending.copy(
            url = result.imageUrl,
            assetId = result.assetId,
            size = result.size,
            status = ImageGenerationStatus.READY,
            userErrorMessage = null,
        )
        val nextHistory = history + HistoryEntry(
            id = ready.id,
            mode = ready.mode,
            provider = ready.provider,
            prompt = ready.prompt,
            assetId = result.assetId,
            assetPath = result.assetPath,
            parentId = if (ready.mode == EditorMode.EDIT) activeParentHistoryId else null,
            createdAt = result.createdAt,
            timestamp = result.timestamp,
            generationBatchId = ready.generationBatchId,
            batchIndex = ready.batchIndex,
        )
        return copy(
            images = images.map { if (it.id == pending.id) ready else it },
            history = nextHistory,
            focusedImageId = ready.id,
            activeRequestId = null,
            activeParentHistoryId = null,
            activePreviousFocusedImageId = null,
            userErrorMessage = null,
        )
    }

    fun failing(requestId: String, message: String): ProviderPipelineState {
        if (activeRequestId != requestId) return this
        val nextImages = images.filterNot { it.id == pendingImageId(requestId) && it.status == ImageGenerationStatus.PENDING }
        return copy(images = nextImages, focusedImageId = restoredFocus(nextImages), activeRequestId = null, activeParentHistoryId = null, activePreviousFocusedImageId = null, userErrorMessage = message)
    }

    fun canceling(requestId: String): ProviderPipelineState {
        if (activeRequestId != requestId) return this
        val nextImages = images.filterNot { it.id == pendingImageId(requestId) && it.status == ImageGenerationStatus.PENDING }
        return copy(images = nextImages, focusedImageId = restoredFocus(nextImages), activeRequestId = null, activeParentHistoryId = null, activePreviousFocusedImageId = null, userErrorMessage = null)
    }

    fun movingToBackground(): ProviderPipelineState {
        val requestId = activeRequestId ?: return copy(lifecyclePhase = AppLifecyclePhase.BACKGROUND)
        return canceling(requestId).copy(lifecyclePhase = AppLifecyclePhase.BACKGROUND)
    }

    fun returningToForeground(network: NetworkReachability): ProviderPipelineState = copy(
        lifecyclePhase = AppLifecyclePhase.FOREGROUND,
        userErrorMessage = if (network == NetworkReachability.OFFLINE) AdapterError.Offline.userMessage else userErrorMessage,
    )

    fun focusing(imageId: String): ProviderPipelineState =
        if (readyImages.any { it.id == imageId }) copy(focusedImageId = imageId) else this

    fun updatingFocusedAnnotations(annotations: CanvasAnnotations): ProviderPipelineState {
        val focused = focusedImageId ?: return this
        return copy(images = images.map { image -> if (image.id == focused) image.copy(annotations = annotations) else image })
    }

    fun deletingHistoryBranch(entryId: String): ProviderPipelineState {
        val nextHistoryState = NativeHistoryBrowserState(history, focusedHistoryEntry?.id).deleting(entryId)
        val remainingIds = nextHistoryState.entries.mapTo(mutableSetOf()) { it.id }
        val remainingAssetIds = nextHistoryState.entries.mapTo(mutableSetOf()) { it.assetId }
        val deletedHistoryIds = history.mapTo(mutableSetOf()) { it.id }.apply { removeAll(remainingIds) }
        val activePendingId = activeRequestId?.let(::pendingImageId)
        val activePending = activePendingId?.let { pendingId -> images.firstOrNull { it.id == pendingId } }
        val activeRequestDeleted = activePendingId == entryId ||
            activeParentHistoryId in deletedHistoryIds ||
            activePending?.parentId in deletedHistoryIds
        val nextImages = images.filter { image ->
            when {
                image.status == ImageGenerationStatus.PENDING -> !activeRequestDeleted || image.id != activePendingId
                else -> image.id in remainingIds || image.assetId in remainingAssetIds
            }
        }
        val selectedImageId = nextHistoryState.selectedEntry?.let { selected ->
            nextImages.firstOrNull { image -> image.id == selected.id || image.assetId == selected.assetId }?.id
        }
        val nextFocus = focusedImageId?.takeIf { focused -> nextImages.any { it.id == focused } }
            ?: selectedImageId
            ?: nextImages.lastOrNull { it.status == ImageGenerationStatus.READY }?.id
        return copy(
            images = nextImages,
            history = nextHistoryState.entries,
            focusedImageId = nextFocus,
            activeRequestId = activeRequestId.takeUnless { activeRequestDeleted },
            activeParentHistoryId = activeParentHistoryId.takeUnless { activeRequestDeleted },
            activePreviousFocusedImageId = activePreviousFocusedImageId.takeUnless { activeRequestDeleted },
        )
    }

    fun movingFocusLeft(): ProviderPipelineState = movingFocus(LineageDirection.LEFT)

    fun movingFocusRight(): ProviderPipelineState = movingFocus(LineageDirection.RIGHT)

    fun movingFocusUp(): ProviderPipelineState = movingFocus(LineageDirection.UP)

    fun movingFocusDown(): ProviderPipelineState = movingFocus(LineageDirection.DOWN)

    fun requestForActivePrompt(
        outputSize: OutputSize = OutputSize.SQUARE,
        references: List<ComposerReferenceSummary> = emptyList(),
        inputImagePath: Path? = null,
        maskImagePath: Path? = null,
    ): ProviderRequest? {
        val requestId = activeRequestId ?: return null
        val pending = images.firstOrNull { it.id == pendingImageId(requestId) } ?: return null
        return ProviderRequest(
            id = requestId,
            prompt = pending.prompt,
            provider = pending.provider,
            mode = pending.mode,
            outputSize = outputSize,
            parentImageId = pending.parentId,
            parentHistoryId = activeParentHistoryId,
            annotations = pending.annotations,
            references = references.map { ProviderReferenceImage(id = it.id, assetPath = it.assetPath) },
            inputImagePath = inputImagePath,
            maskImagePath = maskImagePath,
        )
    }

    private fun focusedReadyImage(): CanvasImage? {
        val focused = focusedImageId ?: return null
        return images.firstOrNull { it.id == focused && it.status == ImageGenerationStatus.READY }
    }

    private fun restoredFocus(nextImages: List<CanvasImage>): String? =
        activePreviousFocusedImageId?.takeIf { previous -> nextImages.any { it.id == previous && it.status == ImageGenerationStatus.READY } }
            ?: nextImages.lastOrNull { it.status == ImageGenerationStatus.READY }?.id

    private fun pendingImageId(requestId: String): String = "pending-$requestId"

    private fun withError(message: String): ProviderPipelineState = copy(userErrorMessage = message)

    private fun placeholderImage(
        id: String,
        requestId: String,
        prompt: String,
        provider: EditorProvider,
        mode: EditorMode,
        parentId: String?,
        generationIndex: Int,
        annotations: CanvasAnnotations = CanvasAnnotations.Empty,
        generationBatchId: String?,
        batchIndex: Int?,
    ): CanvasImage = CanvasImage(
        id = id,
        url = "mock://pending/$requestId",
        assetId = null,
        size = EditorSize(width = 1024.0, height = 1024.0),
        position = EditorPoint(x = 0.0, y = 0.0),
        parentId = parentId,
        generationIndex = generationIndex,
        prompt = prompt,
        provider = provider,
        mode = mode,
        createdAt = generationIndex.toDouble(),
        annotations = annotations,
        hasMagicLayerFields = false,
        status = ImageGenerationStatus.PENDING,
        userErrorMessage = null,
        generationBatchId = generationBatchId,
        batchIndex = batchIndex,
    )

}

File: apps/ios/BananaTape/BananaTape/ProviderPipelineState.swift
import Foundation

enum HistoryDeletionCoordinator {
    static func deleting(entryID: String, from state: ProviderPipelineState) -> ProviderPipelineState {
        state.reconcilingHistory(state.historyBrowserState.deleting(entryId: entryID))
    }
}

struct ProviderPipelineState: Equatable {
    enum LifecyclePhase: Equatable {
        case foreground
        case background
    }

    let images: [CanvasImage]
    let history: [HistoryEntry]
    let focusedImageId: String?
    let activeRequestId: String?
    let activeParentHistoryId: String?
    let userErrorMessage: String?
    let lifecyclePhase: LifecyclePhase

    init(images: [CanvasImage] = [], history: [HistoryEntry] = [], focusedImageId: String? = nil, activeRequestId: String? = nil, activeParentHistoryId: String? = nil, userErrorMessage: String? = nil, lifecyclePhase: LifecyclePhase = .foreground) {
        self.images = images
        self.history = history
        self.focusedImageId = images.first(where: { $0.id == focusedImageId && $0.status == .ready })?.id
            ?? images.last(where: { $0.status == .ready })?.id
        self.activeRequestId = activeRequestId
        self.activeParentHistoryId = activeParentHistoryId
        self.userErrorMessage = userErrorMessage
        self.lifecyclePhase = lifecyclePhase
    }

    var pendingImages: [CanvasImage] { images.filter { $0.status == .pending } }
    var readyImages: [CanvasImage] { images.filter { $0.status == .ready } }
    var focusedImage: CanvasImage? { images.first { $0.id == focusedImageId } }
    var focusedHistoryEntryId: String? {
        guard let focusedImage else { return nil }
        return history.last(where: { $0.id == focusedImage.id || $0.assetId == focusedImage.assetId })?.id
    }
    var historyBrowserState: NativeHistoryBrowserState { NativeHistoryBrowserState(entries: history, selectedEntryId: focusedHistoryEntryId) }
    var lineageAvailability: LineageNavigationAvailability { LineageNavigation.availability(images: images, focusedImageID: focusedImageId) }

    func movingFocus(_ direction: LineageNavigationDirection) -> ProviderPipelineState {
        guard let target = LineageNavigation.target(direction, images: images, focusedImageID: focusedImageId) else { return self }
        return focusing(imageId: target.id)
    }

    func focusing(imageId: String) -> ProviderPipelineState {
        guard images.contains(where: { $0.id == imageId && $0.status == .ready }) else { return self }
        return ProviderPipelineState(images: images, history: history, focusedImageId: imageId, activeRequestId: activeRequestId, activeParentHistoryId: activeParentHistoryId, userErrorMessage: userErrorMessage, lifecyclePhase: lifecyclePhase)
    }

    func replacingFocusedAnnotations(_ annotations: CanvasAnnotations) -> ProviderPipelineState {
        guard let focusedImageId else { return self }
        let nextImages = images.map { image in
            guard image.id == focusedImageId else { return image }
            return CanvasImage(id: image.id, url: image.url, assetId: image.assetId, size: image.size, position: image.position, parentId: image.parentId, generationIndex: image.generationIndex, generationBatchId: image.generationBatchId, batchIndex: image.batchIndex, prompt: image.prompt, provider: image.provider, mode: image.mode, createdAt: image.createdAt, annotations: annotations, hasMagicLayerFields: image.hasMagicLayerFields, status: image.status, userErrorMessage: image.userErrorMessage)
        }
        return ProviderPipelineState(images: nextImages, history: history, focusedImageId: focusedImageId, activeRequestId: activeRequestId, activeParentHistoryId: activeParentHistoryId, userErrorMessage: userErrorMessage, lifecyclePhase: lifecyclePhase)
    }

    func reconcilingHistory(_ browserState: NativeHistoryBrowserState) -> ProviderPipelineState {
        let browserState = browserState.reconcilingLineage()
        let retainedImages = images.filter { image in
            browserState.entries.contains { $0.id == image.id || $0.assetId == image.assetId }
        }
        let selectedImageId = browserState.selectedEntry.flatMap { selectedEntry in
            retainedImages.first { $0.id == selectedEntry.id || $0.assetId == selectedEntry.assetId }?.id
        }
        return ProviderPipelineState(images: retainedImages, history: browserState.entries, focusedImageId: selectedImageId, userErrorMessage: userErrorMessage, lifecyclePhase: lifecyclePhase)
    }

    func startingSubmission(mode: EditorMode, prompt: String, annotations: CanvasAnnotations, requestId: String, network: NetworkReachability, provider: EditorProvider = .mock) -> ProviderPipelineState {
        switch mode {
        case .edit:
            return startingEdit(prompt: prompt, annotations: annotations, requestId: requestId, network: network, provider: provider)
        case .generate:
            return startingGenerate(prompt: prompt, requestId: requestId, network: network, provider: provider)
        }
    }

    func startingGenerate(prompt: String, requestId: String, network: NetworkReachability, provider: EditorProvider = .mock) -> ProviderPipelineState {
        guard network == .online else { return withError(AdapterError.offline.userMessage) }
        let placeholder = placeholderImage(id: pendingImageId(for: requestId), requestId: requestId, prompt: prompt, provider: provider, mode: .generate, parentId: nil, generationIndex: images.count, generationBatchId: requestId, batchIndex: 0)
        return ProviderPipelineState(images: images + [placeholder], history: history, focusedImageId: focusedImageId, activeRequestId: requestId)
    }

    func startingEdit(prompt: String, annotations: CanvasAnnotations, requestId: String, network: NetworkReachability, provider: EditorProvider = .mock) -> ProviderPipelineState {
        guard network == .online else { return withError(AdapterError.offline.userMessage) }
        guard let parentImage = focusedReadyImage, let parentHistory = history.last(where: { $0.id == parentImage.id || $0.assetId == parentImage.assetId }) else { return withError("Select an image before editing.") }
        let placeholder = placeholderImage(id: pendingImageId(for: requestId), requestId: requestId, prompt: prompt, provider: provider, mode: .edit, parentId: parentImage.id, generationIndex: images.count, generationBatchId: requestId, batchIndex: 0, annotations: annotations)
        return ProviderPipelineState(images: images + [placeholder], history: history, focusedImageId: focusedImageId, activeRequestId: requestId, activeParentHistoryId: parentHistory.id)
    }

    func applying(_ result: ProviderImageResult) -> ProviderPipelineState {
        guard activeRequestId == result.requestId, let pending = images.first(where: { $0.id == pendingImageId(for: result.requestId) && $0.status == .pending }) else { return self }
        let ready = pending.ready(result)
        let entry = HistoryEntry(id: ready.id, mode: ready.mode, provider: ready.provider, prompt: ready.prompt, assetId: result.assetId, assetPath: result.assetPath, parentId: ready.mode == .edit ? activeParentHistoryId : nil, generationBatchId: ready.generationBatchId, batchIndex: ready.batchIndex, createdAt: result.createdAt, timestamp: result.timestamp)
        return ProviderPipelineState(images: images.map { $0.id == pending.id ? ready : $0 }, history: history + [entry], focusedImageId: ready.id, lifecyclePhase: lifecyclePhase)
    }

    func failing(requestId: String, message: String) -> ProviderPipelineState {
        guard activeRequestId == requestId else { return self }
        let nextImages = images.filter { !($0.id == pendingImageId(for: requestId) && $0.status == .pending) }
        let nextFocus = nextImages.contains(where: { $0.id == focusedImageId && $0.status == .ready }) ? focusedImageId : nextImages.last(where: { $0.status == .ready })?.id
        return ProviderPipelineState(images: nextImages, history: history, focusedImageId: nextFocus, userErrorMessage: message)
    }

    func canceling(requestId: String) -> ProviderPipelineState {
        guard activeRequestId == requestId else { return self }
        let nextImages = images.filter { !($0.id == pendingImageId(for: requestId) && $0.status == .pending) }
        let nextFocus = nextImages.contains(where: { $0.id == focusedImageId && $0.status == .ready }) ? focusedImageId : nextImages.last(where: { $0.status == .ready })?.id
        return ProviderPipelineState(images: nextImages, history: history, focusedImageId: nextFocus, lifecyclePhase: lifecyclePhase)
    }

    func movingToBackground() -> ProviderPipelineState {
        guard let activeRequestId else {
            return ProviderPipelineState(images: images, history: history, focusedImageId: focusedImageId, userErrorMessage: userErrorMessage, lifecyclePhase: .background)
        }
        let canceled = canceling(requestId: activeRequestId)
        return ProviderPipelineState(images: canceled.images, history: canceled.history, focusedImageId: canceled.focusedImageId, userErrorMessage: canceled.userErrorMessage, lifecyclePhase: .background)
    }

    func returningToForeground(network: NetworkReachability) -> ProviderPipelineState {
        ProviderPipelineState(images: images, history: history, focusedImageId: focusedImageId, userErrorMessage: network == .offline ? AdapterError.offline.userMessage : userErrorMessage, lifecyclePhase: .foreground)
    }

    func requestForActivePrompt(outputSize: OutputSize = .square, references: [ComposerReferenceSummary] = [], inputImageURL: URL? = nil, maskImageURL: URL? = nil) -> ProviderRequest? {
        guard let activeRequestId, let pending = images.first(where: { $0.id == pendingImageId(for: activeRequestId) }) else { return nil }
        return ProviderRequest(
            id: activeRequestId,
            prompt: pending.prompt,
            provider: pending.provider,
            mode: pending.mode,
            outputSize: outputSize,
            parentImageId: pending.parentId,
            parentHistoryId: activeParentHistoryId,
            annotations: pending.annotations,
            references: references.map { ProviderReferenceImage(id: $0.id, assetPath: $0.assetPath) },
            inputImageURL: inputImageURL,
            maskImageURL: maskImageURL
        )
    }

    private var focusedReadyImage: CanvasImage? {
        guard let focusedImageId else { return nil }
        return images.first { $0.id == focusedImageId && $0.status == .ready }
    }

    private func pendingImageId(for requestId: String) -> String { "pending-\(requestId)" }
    private func withError(_ message: String) -> ProviderPipelineState { ProviderPipelineState(images: images, history: history, focusedImageId: focusedImageId, activeRequestId: activeRequestId, activeParentHistoryId: activeParentHistoryId, userErrorMessage: message, lifecyclePhase: lifecyclePhase) }
    private func placeholderImage(id: String, requestId: String, prompt: String, provider: EditorProvider, mode: EditorMode, parentId: String?, generationIndex: Int, generationBatchId: String?, batchIndex: Int?, annotations: CanvasAnnotations = .empty) -> CanvasImage {
        CanvasImage(id: id, url: "mock://pending/\(requestId)", assetId: nil, size: EditorSize(width: 1024, height: 1024), position: EditorPoint(x: 0, y: 0), parentId: parentId, generationIndex: generationIndex, generationBatchId: generationBatchId, batchIndex: batchIndex, prompt: prompt, provider: provider, mode: mode, createdAt: Double(generationIndex), annotations: annotations, hasMagicLayerFields: false, status: .pending, userErrorMessage: nil)
    }
}

private extension CanvasImage {
    func ready(_ result: ProviderImageResult) -> CanvasImage {
        CanvasImage(id: id, url: result.imageURL, assetId: result.assetId, size: result.size, position: position, parentId: parentId, generationIndex: generationIndex, generationBatchId: generationBatchId, batchIndex: batchIndex, prompt: prompt, provider: provider, mode: mode, createdAt: createdAt, annotations: annotations, hasMagicLayerFields: false, status: .ready, userErrorMessage: nil)
    }
}
